File: bayes-align/c-src/align.c
#include <stdio.h>
#include <math.h>
#include <string.h>

#define MATCH 0
#define DIR_I 1
#define DIR_J 2
#define END   3

// RT enzyme error plus the DNA polyermase enzyme error (per base) time the number of cycles of PCR
const double probPCRError = 0.00001490711984999862 + (0.00001038461538461538 * 30.0);

// takes a PHRED+33 encoded quality score from the sequencer and returns the probability of a
// sequencer miscall, as a double
double phredToProbIncorrect(char p)
{
	int phred = (int)p + 33;
	double prob = pow(10, (-((double)phred) / 10));
	return prob;
}

int max_match(double values[3])
{
	int maxIndex = 0;
    for (int i = 1; i < 3; i++)
    {
        if (values[i] > values[maxIndex])
        {
            maxIndex = i;
        }
    }

    return maxIndex;
}

double match(char s, char q, char qual)
{
	double probMisscall = phredToProbIncorrect(qual);

	double probCorrcall = 1.0 - probMisscall;

	double value = 0;

    if (q == 'N')
    {
        value = 1;
    }
	else if (s == q)
	{
		value = (probPCRError * probMisscall) + (probCorrcall * (1.0 - probPCRError));
	}
	else
	{
		value = ((double)1 / 3 * (1.0 - probMisscall)) + ((double)1 / 3 * probCorrcall * probPCRError) + ((double)2 / 9 * probMisscall * probPCRError);
	}
    printf("value: %f\n", value);
	return value;
}

double chance(char s, char q, char qual)
{
	if (s == q)
	{
		return (double)1 / 4;
	}
	else if (s != q && q != 'N')
	{
		return (double)3 / 4;
	}
	else
	{
		return (double)1;
	}
}


void printMatrix(int rows, int cols, double m[rows][cols])
{
	for (int i = 0; i<rows; i++)
	{
		for (int j = 0; j<cols; j++)
		{
			printf("%2.3f\t", m[i][j]);
		}
		printf("\n");
	}
	printf("\n");

}


char* subject = "GTGTCAGTCACTTCC";

char* query = "NTGTGCANNNNCTTCC";

// var query = []byte("GTGTCAGTCACTTCC")

char* quals = "@#-%@EC####-,;BC";

void align(char* subj, char* query, char* quals )
{

    int max_i = strlen(subj)+1;
    int max_j = strlen(query)+1;

    double matrixRatio[max_i][max_j];

    double matrixSL[max_i][max_j];
    double matrixSLP[max_i][max_j];
    double matrixDir[max_i][max_j];

    //fill the matrix

	for (int i = 0; i < max_i; i++)
	{
		for (int j = 0; j < max_j; j++)
		{
			if (i == 0 || j == 0)
			{
				matrixRatio[i][j] = 1.0;

                matrixSL[i][j] = 1.0;
                matrixSLP[i][j] = 1.0;
                matrixDir[i][j] = END;

			}
			else
			{
                double potentialMoves[3] =
                {
                    matrixSL[i-1][j-1]*match(query[j-1], subj[i-1], quals[j-1]),
                    matrixSL[i-1][j]*probPCRError,
                    matrixSL[i][j-1]*probPCRError
                };
                printf("%f, %f, %f\n", potentialMoves[0], potentialMoves[1], potentialMoves[2]);

				int bestMoveIndex = max_match(potentialMoves);

                int dir = 0;
                int lastDir_i = 0;
                int lastDir_j = 0;
                switch(bestMoveIndex)
                {
                    case 0:
                        dir = MATCH;
                        lastDir_i = i-1;
                        lastDir_j = j-1;
                        break;
                    case 1:
                        dir = DIR_I;
                        lastDir_i = i-1;
                        lastDir_j = j;
                        break;
                    case 2:
                        dir = DIR_J;
                        lastDir_i = i;
                        lastDir_j = j-1;
                        break;
                }

                double probSL = potentialMoves[bestMoveIndex];

				double probSLP = chance(query[lastDir_j], subj[lastDir_i], quals[lastDir_j])*matrixSLP[lastDir_i][lastDir_j];

				matrixRatio[i][j] = probSL / probSLP;

                matrixSL[i][j] = probSL;
                matrixSLP[i][j] = probSLP;
                matrixDir[i][j] = dir;
			}
		}
	}

	printMatrix(max_i, max_j, matrixRatio);
    printf("dir matrix:\n");
    printMatrix(max_i, max_j, matrixDir);

}

int main(void)
{
	align(subject, query, quals);
}

File: bayes-align/align.go
package main

import (
	"fmt"
	"math"
)

var probPCRError = 0.00001490711984999862 + 0.00001038461538461538*30

func phredToProbIncorrect(p byte) float64 {
	phred := int(p) + 33
	prob := math.Pow(10, (-float64(phred) / 10))
	return prob
}

func max(d, i, j float64) (float64, int) {
	max := d
	dir := 1

	if i > max {
		max = i
		dir = 2
	}
	if j > max {
		max = j
		dir = 3
	}
	// fmt.Println(m)
	// fmt.Println(max)
	return max, dir
}

func match(s, q, qual byte) float64 {
	probMisscall := phredToProbIncorrect(qual)

	probCorrcall := 1 - probMisscall

	var value float64
	if s == q {
		// fmt.Println(probPCRError * probMisscall)
		value = (probPCRError * probMisscall) + (probCorrcall * (1 - probPCRError))
	} else {
		value = (float64(1) / 3 * (1 - probMisscall)) + (float64(1) / 3 * probCorrcall * probPCRError) + (float64(2) / 9 * probMisscall * probPCRError)
	}
	return value
}

func chance(s, q, qual byte) float64 {
	if s == q {
		return float64(1) / 4
	} else if s != q && q != 'N' {
		return float64(3) / 4
	} else {
		return 1
	}
}

var subject = []byte("GTGTCAGTCACTTCC")

var query = []byte("NTGTGCANNNNCTTCC")

// var query = []byte("GTGTCAGTCACTTCC")

var quals = []byte("@#-%@EC####-,;BC")

func align(subj, query, quals []byte) [][]float64 {

	matrix := make([][]float64, len(subj)+1)

	matrixSL := make([][]float64, len(subj)+1)

	matrixSLP := make([][]float64, len(subj)+1)

	matrixDir := make([][]int, len(subj)+1)

	for i := range matrix {
		matrix[i] = make([]float64, len(query)+1)
		matrixSL[i] = make([]float64, len(query)+1)
		matrixSLP[i] = make([]float64, len(query)+1)
		matrixDir[i] = make([]int, len(query)+1)
	}

	for i := range matrix {
		for j := range matrix[i] {
			if i == 0 || j == 0 {
				matrix[i][j] = 1

				// matrixSL[i][j] =
				// matrixSLP[i][j] =
				// matrixDir[i][j] =

			} else if query[j-1] == byte('N') {
				matrix[i][j] = matrix[i-1][j-1]

				// matrixSL[i][j] =
				// matrixSLP[i][j] =
				// matrixDir[i][j] =

			} else {
				probSC, dir := max(
					matrix[i-1][j-1]*match(query[j-1], subj[i-1], quals[j-1]),
					matrix[i-1][j]*probPCRError,
					matrix[i][j-1]*probPCRError,
				)
				if dir == 1 {

				}
				probSCP := chance(query[j-1], subj[i-1], quals[j-1])

				// TODO: this is wrong: we need to store each of these values individually... and
				// recalculated at each one... the alignment matrix could be probSC/probSCP
				matrix[i][j] = probSC / probSCP

				// matrixSL[i][j] =
				// matrixSLP[i][j] =
				// matrixDir[i][j] =

			}
		}
	}

	// find highest SL/SLP ratio

	// traceback

	// return alignment representation, and P(S|L), and P(S|L')

	return matrix

}

//
// func isAlignmentLinker(a alignment) bool {

// take alignment, calculate P(L|S) and P(L'|S) and compare, return true/false
// }

func printMatrix(m [][]float64) {
	for i := range m {
		for j := range m[i] {
			fmt.Printf("%2.3f\t", m[i][j])
		}
		fmt.Printf("\n")
	}
	fmt.Printf("\n")

}

func main() {
	m := align(subject, query, quals)
	printMatrix(m)
}
